randovania/bitpacking/construct_pack.py

```diff
@@ -13,17 +13,32 @@
 
 from randovania.lib import type_lib
 
+if typing.TYPE_CHECKING:
+    from _typeshed import DataclassInstance
+
 BinStr = construct.PascalString(construct.VarInt, "utf-8")
 
 
 T = typing.TypeVar("T")
 
+JsonValue = str | int | float | list["JsonValue"] | dict[str, "JsonValue"]
+
+
+@typing.runtime_checkable
+class JsonEncodable(typing.Protocol):
+    @classmethod
+    def from_json(cls, value: JsonValue) -> typing.Self:
+        ...
```

Comment:
What do these 3 dots do?

Reply:
nothing! it just indicates that children are expected to implement this function (consider it abstract)

Reply:
I copied from typing.Protocol documentation, saw it as "dummy function thing" and seems to be it. Looking it up, called `Ellipsis` and seems to be a fancy way of doing `pass`?

Reply:
if it's the same thing as pass, then i'd rather just use pass.

Reply:
it's functionally equivalent to `pass`, but semantically it means "this must be implemented in the children" without explicitly raising a NotImplementedError

Reply:
ah ok

Reply:
upon further inspection: it's being used here because this Protocol is serving as a sort of formalized duck typing. by using an ellipse, the meaning becomes more clear: any object with methods of this name/signature is considered to be this type, regardless of the contents of those methods. using `pass` may give the wrong impression that these methods must be no-op. ALSO, I'd bet that mypy would be unhappy about the functions not returning their assigned types, but has special handling for the ellipse here